docs/kb/semgrep-ci/azure-self-hosted-ubuntu.md

```diff
@@ -0,0 +1,131 @@
+---
+tags:
+  - Azure Pipelines
+description: Run Semgrep on self-hosted Ubuntu runners in Azure DevOps.
+---
+# Semgrep with self-hosted Ubuntu runners in Azure Pipelines
+
+Semgrep provides a [sample configuration for Azure-hosted runners](/docs/semgrep-ci/sample-ci-configs#azure-pipelines). If you use self-hosted Ubuntu Linux runners, you have significantly more control over their configuration, but as a result, they require additional preparation and configuration to run Semgrep.
+
+This guide adds two approaches to configuring self-hosted runners that use Ubuntu (the default self-hosted option for Azure DevOps Linux runners):
+
+* [Using pipx](#using-pipx)
+* [Using pip with a virtual environment](#using-pip-with-a-virtual-environment)
+
+## Using pipx
+
+While the sample configuration uses `pip`, this approach uses `pipx`, which avoids issues with system-managed Python vs user-installed Python.
+
+### Prepare your runner
+
+Access the runner and execute the following commands:
+
+```bash
+$ sudo apt update
+$ sudo apt install pipx
+$ pipx ensurepath
+```
+
+After completing the commands:
+
+1. Start a new shell session, so that the changes from `pipx ensurepath` are available.
+2. Ensure the [Azure DevOps agent](https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/linux-agent?view=azure-devops) is set up and running.
+
+### Create your configuration
+
+1. Follow the steps provided in the [sample configuration for Azure-hosted runners](/docs/semgrep-ci/sample-ci-configs#azure-pipelines).
+2. Add the following snippet to the `azure-pipelines.yml` for the repository.
+
+```yaml
+variables:
+- group: Semgrep_Variables
+
+pool:
+  name: Default
+
+steps:
+- checkout: self
+  clean: true
+  fetchDepth: 20
+  persistCredentials: true
+- script: |
+    pipx install semgrep
+    if [ $(Build.SourceBranchName) = "master" ]; then
+        echo "Semgrep full scan"
+        semgrep ci
+    elif [ $(System.PullRequest.PullRequestId) -ge 0 ]; then
+        echo "Semgrep diff scan"
+        git fetch origin master:origin/master
+        export SEMGREP_PR_ID=$(System.PullRequest.PullRequestId)
+        export SEMGREP_BASELINE_REF='origin/master'
+        semgrep ci
+    fi
+  env:
+    SEMGREP_APP_TOKEN: $(SEMGREP_APP_TOKEN)
+```
+
```

Comment:
This is basically where the problem is (and the equivalent place for the second config) - in order for these configs to work, the user would need to also follow the instructions on line 29+ of `src/components/code_snippets/_azure-semgrep-app-sast.mdx`, but it's difficult to refer to on its own.

Reply:
Maybe we should also snippetify those instructions so that the user doesn't have to leave the KB article?

Reply:
@s-santillan I gave this a shot, if you don't mind taking another look! Sorry for the delay, and I hope I put the snippet in the right place.

Reply:
I checked the [preview](https://deploy-preview-1887--semgrep-docs-prod.netlify.app/docs/kb/semgrep-ci/azure-self-hosted-ubuntu) and it looks good! But in the sample CI configs looks like there's a [duplicate](https://deploy-preview-1887--semgrep-docs-prod.netlify.app/docs/semgrep-ci/sample-ci-configs#setting-environment-variables-in-azure-pipelines)?

Reply:
Oh, thank you. I probably just forgot to remove the old one when I added the snippet, I'll get that fixed up.

Reply:
Should be fixed now!